fix(ssi): decouple admission controller webhooks

betterengineering opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
This change refactors three mutating webhooks (`config`, `tagsfromlabels`, and `autoinstrumentation`) to support a common `Mutator` interface along with a common `MutationFilter` interface. This allows the logic from each to be decoupled from the underlying webhook so that the logic can be reused. Additionally, the `autoinsrumentation` webhook will now utilize the mutation logic for all three mutators.

BREAKING CHANGE: As part of this, we've added enabled/disabled namespace configuration options to the `config` and `tagsfromlabels` webhooks and they will no longer rely on the `apm_config.instrumentation` based config options used for SSI.

### Motivation
Without this change, we have three webooks that are tightly coupled: `config`, `tagsfromlabels`, and `autoinstrumentation`. All three webhooks require config values from `autoinstrumentation` for namespace filtering. Additionally, `autoinstrumentation` requires the other two webhooks to be enabled to function properly. 

As part of  [Kubernetes SSI | Workload Selection 🎯](https://docs.google.com/document/d/1Lol1ExLF1nGBe7njO6DBVkjuYAYxC1-sL2WP0rdqFRk/edit?usp=sharing), we want to substantially modify the way we do filtering

Decoupling these webhooks is a necessary step to safely modifying the single step configuration. See [Decoupling Admission Controller Webhooks](https://docs.google.com/document/d/1emhPd8kWMrGELKuGem1ERMFY5IPZOx_nDR7oRoTvQvA/edit?usp=sharing) for more details.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

I tested the following scenarios to ensure that each of the webhooks had the expected behavior:
* Default Settings with SSI Enabled
* SSI Enabled and Targets and Config Disabled
* SSI Disabled and Targets and Config Enabled

See [#33599 Testing Plan](https://docs.google.com/document/d/1sdt0vRLIgc86-B64tldel1JxVW35IZUCzfpJMW_x_AM/edit?usp=sharing) for more details.

### Possible Drawbacks / Trade-offs
The most obvious draw back is that we are introducing a breaking change with how these webhooks are configured so that they are all using dedicated configurations. We believe this is both the correct solution moving forward and will offer the best user experience. For example, customers who manually configure their tracers would still want things like the config webhook to work even if they are not using SSI. Today, that is not possible as it requires SSI to be enabled.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
This change is quite large. If there are nice to haves, let me know and I'd be more than happy to follow up with additional changes.

betterengineering committed:
fix(ssi): migrate injection filter to mutate common

This commit migrates the injection filter into the mutate/common package
given it's used by multiple webhooks.

betterengineering committed:
fix(ssi): remove config from injection filter

This commit moves config fetching out of the NewInjectorFIlter method
and instead requires it to be passed in.

betterengineering committed:
fix(ssi): update mutating webhooks to use their own config

BREAKING CHANGE: This commit updates the config webhook and the tags to
labels webhook to use their own configs when constructing their
injection filter. While there were no substantial changes to the tests,
this change could certainly have subtle consequences.

betterengineering committed:
Fixup injection_filter

betterengineering committed:
Fixup namespace filter naming

betterengineering committed:
Fixup log line

adel121 commented on pkg/clusteragent/admission/mutate/common/injection_filter.go:
```diff
@@ -73,7 +66,7 @@ func (f *injectionFilter) Err() error {
 }
 
-// makeAPMSSINamespaceFilter returns the filter used by APM SSI to filter namespaces.
+// makeNamespaceFilter returns the filter used by APM SSI to filter namespaces.
```
> used by APM SSI

this function is no longer specific to the autoinstrumentation webhook

adel121 commented on pkg/clusteragent/admission/mutate/tagsfromlabels/tags.go:
```diff
@@ -56,10 +56,15 @@ type Webhook struct {
 // NewWebhook returns a new Webhook
-func NewWebhook(wmeta workloadmeta.Component, datadogConfig config.Component, injectionFilter mutatecommon.InjectionFilter) *Webhook {
+func NewWebhook(wmeta workloadmeta.Component, datadogConfig config.Component) *Webhook {
+	enabled := datadogConfig.GetBool("admission_controller.inject_tags.enabled.enabled")
```
```suggestion
	enabled := datadogConfig.GetBool("admission_controller.inject_tags.enabled")
```

adel121 commented on pkg/clusteragent/admission/mutate/tagsfromlabels/tags.go:
```diff
@@ -56,10 +56,15 @@ type Webhook struct {
-func NewWebhook(wmeta workloadmeta.Component, datadogConfig config.Component, injectionFilter mutatecommon.InjectionFilter) *Webhook {
+func NewWebhook(wmeta workloadmeta.Component, datadogConfig config.Component) *Webhook {
+	enabled := datadogConfig.GetBool("admission_controller.inject_tags.enabled.enabled")
+	enabledNamespaces := datadogConfig.GetStringSlice("admission_controller.inject_tags.enabled_namespaces")
+	disabledNamespaces := datadogConfig.GetStringSlice("admission_controller.inject_tags.disabled_namespaces")
```
These 2 config keys are new, they should be declared in `pkg/config/setup` [here](https://github.com/DataDog/datadog-agent/blob/cd041a215b45cc3fe9f5bf8064d76b5f56292edb/pkg/config/setup/config.go#L758).

You can find an example for auto instrumentation webhook configuration [here](https://github.com/DataDog/datadog-agent/blob/cd041a215b45cc3fe9f5bf8064d76b5f56292edb/pkg/config/setup/apm.go#L85-L86).

adel121 commented on pkg/clusteragent/admission/mutate/config/config.go:
```diff
@@ -86,10 +86,15 @@ type Webhook struct {
-func NewWebhook(wmeta workloadmeta.Component, injectionFilter mutatecommon.InjectionFilter, datadogConfig config.Component) *Webhook {
+func NewWebhook(wmeta workloadmeta.Component, datadogConfig config.Component) *Webhook {
+	enabled := datadogConfig.GetBool("admission_controller.inject_config.enabled")
+	enabledNamespaces := datadogConfig.GetStringSlice("admission_controller.inject_config.enabled_namespaces")
+	disabledNamespaces := datadogConfig.GetStringSlice("admission_controller.inject_config.disabled_namespaces")
```
Same comment as for tagslabels 

These config keys are new and should be declared [here](https://github.com/DataDog/datadog-agent/blob/cd041a215b45cc3fe9f5bf8064d76b5f56292edb/pkg/config/setup/config.go#L750).

betterengineering committed:
Update pkg/clusteragent/admission/mutate/tagsfromlabels/tags.go

Co-authored-by: Adel Haj Hassan <41540817+adel121@users.noreply.github.com>

betterengineering committed:
Address PR feedback

betterengineering committed:
fix(ssi): create injector interface and refactor webhooks

This commit adds an injector interface and refactors the tagsfromlabels
and the config webhooks to use it. This will allow us to use the
injector logic without needing to instantiate a webhook.

betterengineering committed:
fix(ssi): rename internal instruementation pod injector

This commit renames the injector struct in the autoinstrumentation
package so that we can reuse the name/pattern from the other webhooks.

stanistan commented on pkg/clusteragent/admission/mutate/config/injector.go:
```diff
@@ -0,0 +1,248 @@
+//   - DD_AGENT_HOST: the host IP of the node
+//   - DD_ENTITY_ID: the entity ID of the pod
+//   - DD_EXTERNAL_ENV: the External Data Environment Variable
+func (i *ConfigInjector) InjectPod(pod *corev1.Pod, ns string, dc dynamic.Interface) (bool, error) {
```
```go
type injectorMutatorBetterNameAdapter struct {
   injector Injector
}

func (i *injectorMutatorBetterNameAdapter) MutatePod(pod *corev1.Pod) error {
   // fix pod namespace with ns string
   _, err := i.injector.InjectPod(pod)
}
```

betterengineering committed:
fix(ssi): update auto instrumentation to new interface

This commit converts the auto instrumentation webhook to have a
seperated injector and webhook logic.

betterengineering committed:
fix(ssi): refactor to mutator and mutation filter

This commit refactors to use a mutator and a mutation filter.
Additionally, it refactors the instantiation inside of the controller
base.

betterengineering committed:
fix(ssi): cleanup mutation filter interface

This commit cleans up the common and default mutation filter so it's
more readable and concise.

betterengineering committed:
fix(ssi): add a multi mutator implementation

This commit adds an multi mutator implementation that will treat a list
of mutators as a single mutation.

betterengineering committed:
fix(ssi): wire in multi mutator for apm

This commit wires in the multi mutator implementation for apm so that
all three mutators apply.

betterengineering committed:
Fix label selector issue

betterengineering committed:
Address PR feedback

betterengineering committed:
Add release note

betterengineering committed:
Merge branch 'main' into mark.spicer/INPLAT-467-seperate-usages-of-injection-filter

drichards-87 reviewed (approved):
Left some small suggestions from Docs and approved the PR.

stanistan commented on pkg/clusteragent/admission/common/label_selectors.go:
```diff
@@ -19,7 +19,11 @@ func DefaultLabelSelectors(useNamespaceSelector bool) (namespaceSelector, object
 
 	if pkgconfigsetup.Datadog().GetBool("admission_controller.mutate_unlabelled") ||
 		pkgconfigsetup.Datadog().GetBool("apm_config.instrumentation.enabled") ||
-		len(pkgconfigsetup.Datadog().GetStringSlice("apm_config.instrumentation.enabled_namespaces")) > 0 {
```
A note for those reading this, this is used across all webhooks and isn't very optimized because of the dependencies across all of them (which we are trying to decouple).

@betterengineering Add a `TODO` here to create different label selectors and use them for the different webhooks (or make one webhook?). There are options. 

cc @davidor @adel121 this definitely needs a follow up from your team.

stanistan commented on pkg/clusteragent/admission/mutate/config/filter.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func NewFilter(cfg *FilterConfig) (mutatecommon.MutationFilter, error) {
+	return mutatecommon.NewDefaulFilter(cfg.enabled, cfg.enabledNamespaces, cfg.disabledNamespaces)
```
function name typo: `NewDefaultFilter`

stanistan commented on pkg/clusteragent/admission/mutate/config/filter.go:
```diff
@@ -0,0 +1,31 @@
+	mutatecommon "github.com/DataDog/datadog-agent/pkg/clusteragent/admission/mutate/common"
+)
+
+type FilterConfig struct {
```
I'm putting down my thoughts on the structure here having some more time to think about the `FilterConfig`... 

- Each of the `Config` structs is currently public with private fields. That means either folks can construct an empty `FilterConfig{}`, or they can use `NewFilterConfig`-- but really there's only one way of constructing them. Are they private for internal testing? We're not doing that though.

- If the fields aren't meant to ever be touched from the outside, why is this function being exposed this way?

--- 

TLDR: maybe its naming? 😭 

- Rename `NewFilter()` -> `MutationFilter` (for consistency with what it does).
- Rename `NewFilterConfig()` -> `NewMutationFilterConfig()`

No pressure to change anything.

stanistan commented on pkg/clusteragent/admission/mutate/common/mutator.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// NewMutators creates a new MultiMutator with the provided Mutators.
+func NewMutators(mutators ...Mutator) *Mutators {
```
find `MultiMutator` and delete the word :)

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go:
```diff
@@ -3584,19 +3599,50 @@ func TestShouldInject(t *testing.T) {
 	}
 }
 
+func mustInjector(t *testing.T, wmeta workloadmeta.Component, ddConfig config.Component) *Mutator {
```
`mustMutator`

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go:
```diff
@@ -1069,13 +1077,17 @@ func TestExtractLibInfo(t *testing.T) {
+			require.NoError(t, err)
+			cfg, err := NewMutatorConfig(mockConfig)
+			require.NoError(t, err)
+			injector := NewMutator(cfg, filter, wmeta)
```
`s/injector/mutator` exceptt for when you're not supposed to 😭

stanistan commented on pkg/clusteragent/admission/controllers/webhook/controller_base.go:
```diff
@@ -118,38 +111,96 @@ func (c *controllerBase) generateWebhooks(wmeta workloadmeta.Component, pa workl
+	}
+	mutatorCfg := configWebhook.NewMutatorConfig(datadogConfig)
+	mutator := configWebhook.NewMutator(mutatorCfg, filter)
+	return configWebhook.NewWebhook(wmeta, datadogConfig, mutator), nil
```
This might be out of scope for this change, but do the `NewWebhok` functions need to have the `datadogConfig` passed in anymore?

stanistan reviewed (approved):
This LGTM! There's a few remnants of things being named weird in the PR but this is a hairy change to get right, nice work.

betterengineering committed:
Apply suggestions from code review

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

davidor reviewed (approved):
Looks good to me.

We can address https://github.com/DataDog/datadog-agent/pull/33599/files#r1942087592 in a separate PR.

I'm still not sure if we need to add the 4 new options to support namespace filtering in the config and tagsfromlabels webhooks, but we can change that if needed before the next release. This is something that was not really supported before, so I don't think we need to add it now. If we need to add this feature we don't have to do it by namespace name.

hush-hush commented on pkg/config/setup/config.go:
```diff
@@ -756,6 +758,8 @@ func InitConfig(config pkgconfigmodel.Setup) {
 	config.BindEnvAndSetDefault("admission_controller.inject_config.type_socket_volumes", false)
 	config.BindEnvAndSetDefault("admission_controller.inject_tags.enabled", true)
+	config.BindEnvAndSetDefault("admission_controller.inject_tags.enabled_namespaces", []string{})
+	config.BindEnvAndSetDefault("admission_controller.inject_tags.disabled_namespaces", []string{})
```
Should we add those 4 new settings to the configuration template (`pkg/config/config_template.yaml`) ?

betterengineering replied:
We're still debating if we need to actually expose these options or not. I added this comment as a follow up to make sure it's not dropped: https://datadoghq.atlassian.net/browse/INPLAT-472

hush-hush reviewed (approved):
Added a quick question but looks good for agent-configuration

betterengineering commented:
/merge

betterengineering commented:
/remove

betterengineering committed:
Fix linter issues

betterengineering commented:
/merge

betterengineering committed:
Fix tests

betterengineering committed:
Merge branch 'main' into mark.spicer/INPLAT-467-seperate-usages-of-injection-filter

betterengineering commented:
/merge
